### ISSUE TITLE
Integer scaling

### ISSUE THREAD

**Issue by chaoskagami** [WayfireWM/wayfire#921]
I play a lot of old games. The windows of these old games are beyond tiny on any sufficiently modern display. Some allow changing the window size, but doing so while keeping the aspect ratio is hard (I used xdotool previously for this), and often they end up with filtering (blur) when ideally there would be none (if it's a pixellated game.)

Since wayland doesn't have a lot of the limitations of X11 with how windows are handled (it's all just composited layers), integer scaling via a transform should be possible.

To be clear, I'm not talking about resizing the window. I'm talking about applying a transformation to render the window larger, preferably with configurable filtering.

(I tried to implement a plugin myself by adding a transformer, but it's crashy and I have no idea how to properly handle damage with this, so I'm opening an issue instead.)

**Comment by soreau** [comment]
If I understand you correctly, I think the [force-fullscreen](https://github.com/WayfireWM/wayfire-plugins-extra/blob/master/metadata/force-fullscreen.xml) plugin might be what you're looking for. It is part of wayfire-plugins-extra.

**Comment by damianatorrpm** [comment]
It would be sweet if a workaround plugin would do this automatically but this would be bad to implement I imagine.
Not all xwayland apps have NET_WM information about their size (but they all should have NET_WM_FULLSCREEN state).
Even with that it can't be certain what the users intend is (another plugin might allow windows to be spanned across all screens) and this would conflict?

**Comment by chaoskagami** [comment]
@soreau No, it's not. This puts the application into a fullscreen mode where the contents are scaled up. I'd ideally like to continue doing other things on the same workspace - this doesn't allow that. It also doesn't have any options for integer scaling or filtering modes in general, so things get billinear filtering which I find undesirable.

@damianatorrpm Dunno about automatically, I was thinking more key mappings to increase scale by one and decrease scale by one. I have no idea if it would conflict across multiple outputs.

**Comment by soreau** [comment]
@chaoskagami Ah, I have had some idea that there could be some binding like <mod>+scroll that could change the size of the view scaled from the center of the view. Would this work?

**Comment by chaoskagami** [comment]
@soreau This request isn't about window size. Some of these older games that are run in emulators, WINE or otherwise do not physically support being resized and will end up with empty borders if you do so.

**Comment by soreau** [comment]
Sorry, I meant apply a transformer that would scale the view, not change its actual size..

**Comment by chaoskagami** [comment]
@soreau If I can change the increment to only use pixel sizes and turn off GL's billinear filtering, sure.

(Though to be honest I prefer key combinations to scroll-based adjustments - working with a touchpad here, unfortunately.)

**Comment by ammen99** [comment]
@chaoskagami Can you share your code where you tried the custom transformer approach? This is definitely the correct way forward.

**Comment by chaoskagami** [comment]
Unfortunately, I already deleted it. The last state I had my terrible attempt in resulted in a black screen and ultimately required a SysRq+REISUB to recover from. I was probably doing something wrong, but the API isn't that well documented and I'm rather new to wayland stuff.

I will say that I used the `alpha` plugin as a reference but adjusted `scale_x`/`scale_y` instead of `alpha`.

**Comment by soreau** [comment]
[This](http://ix.io/2GIu) is a patch for wayfire-plugins-extra.

**Comment by chaoskagami** [comment]
Works nicely here, aside from two quirks:

 * setting the `inc_x_binding`/`inc_y_binding` to the same bind (as well as the same for dec) increases the window scale twice via the keyboard combo (so, to 3x size.) Setting `zoom_step` to 0.5 works around this, but the proper fix is probably ignoring/unsetting one of the keyboard binds if `preserve_aspect` is enabled since `delta.x = delta.y = std::max(delta.x, delta.y);` will cause the X transform to apply to both dimensions and then the Y transform as well.

 * The titlebar is also zoomed if `decoration` or `windecor` is enabled (but honestly, I'm not really sure this one is fixable due to how transforms are implemented)

**Comment by soreau** [comment]
> Works nicely here, aside from three quirks:
> 
>     * Setting the `inc_x_binding`/`inc_y_binding` to the same bind (as well as the same for `dec`) increases the window scale twice via the keyboard combo (so, to 3x size) with `preserve_aspect` enabled. Setting `zoom_step` to 0.5 works around this as does removing one of the binds. The ultimate cause is `delta.x = delta.y = std::max(delta.x, delta.y);` as far as I can tell, it will cause the X transform to apply to both dimensions and then the Y transform as well with `preserve_aspect` since it gets called twice. Might be worth checking if `preserve_aspect` is enabled and immediately returning if processing `y`.

I consider this a user error. If you bind two actions to the same key, both actions will happen.

> 
>     * Any plugin that resizes a `winzoom` window based on output/workspace geometry (maximizing, `tile`, etc) results in wrong x,y-positioning/size. Granted, the point of this is to avoid resizing, but the bad interaction is worth noting.  I don't know what a proper fix for this would be. Disable resize somehow? Make plugins that resize windows aware of `scale_x`/`scale_y` transforms?
> 

I can see what you mean, but the view is scaled by a transformer and this is what happens when you do that. I would consider this plugin only for strict use-cases of making what you originally outlined possible, not much more.

>     * The titlebar is also zoomed if `decoration` or `windecor` is enabled (and honestly, I'm not really sure this one is fixable due to how transforms are implemented if I understand things correctly, so this is more food for thought than anything. I don't use either plugin personally.)
> 

Yes, this is just how it works. You could technically avoid this only for SSD but it would probably be more work than it's worth.

> 
> Regardless, it does fit my use case and I'll be using it. Hopefully that patch can land in wayfire-plugins-extra.

That is good to hear. I think I will add it to wayfire-plugins-extra. Any other feedback is welcome.

**Comment by chaoskagami** [comment]
> I consider this a user error. If you bind two actions to the same key, both actions will happen.

Fair enough. I doubt I'm the last person that will run into that one, though. It's worth at least pointing that caveat out here for anyone searching in the future.

> I can see what you mean, but the view is scaled by a transformer and this is what happens when you do that. I would consider this plugin only for strict use-cases of making what you originally outlined possible, not much more.

Also fair. "Fixing" this would probably require a bunch of work inside the core of wayfire for *very* little payoff, and I doubt the codepath will even end up used 99% of the time.

>Yes, this is just how it works. You could technically avoid this only for SSD but it would probably be more work than it's worth.

For client side decoration, yeah, it's unavoidable. Those are part of the application after all. For SSD, just from reading through that definitely seems like way too much work for too little gain so I completely understand not wanting to do anything.

In any case, thank you very much for the quick work. This is very useful and is (yet another) reason to ditch X11 permanently for me. I suppose whenever it gets added to wayfire-plugins-extra this issue can be closed?
